src/main/java/org/logstash/plugins/inputs/http/NettyHttpServer.java

```diff
@@ -0,0 +1,94 @@
+import io.netty.channel.*;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.nio.NioServerSocketChannel;
+
+import java.io.Closeable;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Created by joaoduarte on 11/10/2017.
+ */
+public class NettyHttpServer implements Runnable, Closeable {
+    private final EventLoopGroup processorGroup =
+            new NioEventLoopGroup(0, daemonThreadFactory("http-input-processor"));
+    private final ServerBootstrap serverBootstrap;
+    private final String host;
+    private final int port;
+    private SslSimpleBuilder sslSimpleBuilder;
+
+    public NettyHttpServer(final String host, final int port, final SslSimpleBuilder sslSimpleBuilder)
+    {
+        this.host = host;
+        this.port = port;
+        this.sslSimpleBuilder = sslSimpleBuilder;
+        serverBootstrap = new ServerBootstrap()
+                .group(processorGroup)
+                .channel(NioServerSocketChannel.class)
+                .childOption(ChannelOption.SO_KEEPALIVE, true);
+    }
+
+    public void childHandler(IMessageHandler childChandler) {
+        final HttpInitializer httpInitializer = new HttpInitializer(childChandler);
+        if (sslSimpleBuilder != null) {
+            httpInitializer.enableSSL(sslSimpleBuilder);
+        }
+        serverBootstrap.childHandler(httpInitializer);
+    }
+
+    public void connectionBacklog(final int connectionBacklog) {
+        serverBootstrap.option(ChannelOption.SO_BACKLOG, connectionBacklog);
+    }
+
+    @Override
+    public void run() {
+        try {
+            final ChannelFuture channel = serverBootstrap.bind(host, port);
+            channel.sync().channel().closeFuture().sync();
+        } catch (final InterruptedException ex) {
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    @Override
+    public void close() {
+        try {
+            processorGroup.shutdownGracefully(0, 10, TimeUnit.SECONDS).sync();
+        } catch (final InterruptedException ex) {
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    public static ThreadFactory daemonThreadFactory(String namePrefix) {
+        return new DaemonThreadFactory(namePrefix);
+    }
+
+    /*
+
+     */
+    static class DaemonThreadFactory implements ThreadFactory {
+
+        final ThreadGroup group;
+        final AtomicInteger threadNumber = new AtomicInteger(1);
+        final String namePrefix;
+
+        DaemonThreadFactory(String namePrefix) {
+            this.namePrefix = namePrefix;
+            group = Thread.currentThread().getThreadGroup();
+        }
+
+        @Override
+        public Thread newThread(Runnable r) {
+            Thread t = new Thread(group, r,
+                    namePrefix + "[T#" + threadNumber.getAndIncrement() + "]",
+                    0);
+            t.setDaemon(true);
```

Comment:
Interesting comment by @original-brownbear in https://github.com/elastic/support-dev-help/issues/2816#issuecomment-347161856:

> @jsvd hmm, (wrong place to discuss this :D), but generally speaking I think it's a big anti-pattern to make I/O threads daemon threads. They pretty much always hold state in their buffers and you want that state to be cleanly resolved.
> If you just make the Daemon threads you have no assurances to that end. If shutdown on interrupting once gets stuck, that means that some thread is simply not joining because the interrupt handling in that thread is wrong.
> Http and beats are a good example here, if you just turn those into daemons, won't you still accept connections from them when the pipeline is already shut down? Seems to me that causes a lot of needless connection reset by peer, when you could just as well kill the server socket first, finish with the client socket handling and then stop the pipeline?

Reply:
> They pretty much always hold state in their buffers and you want that state to be cleanly resolved.

We already control the graceful shutdown sequence by catching sigint and calling the agent.shutdown, which attempts to shutdown everything in a controlled manner, namely doing `processorGroup.shutdownGracefully(0, 10, TimeUnit.SECONDS).sync();` to terminate the nio loop group.

> Http and beats are a good example here, if you just turn those into daemons, won't you still accept connections from them when the pipeline is already shut down?

I'm not sure how is this related to a thread being daemonized or not, this window of accepting connections can be tuned with the grace_period and timeout of that `shutdownGracefully` call

Reply:
@jsvd I admit I didn't read the whole code here and was just making general statements and maybe the implementation here takes care of everything, but:

> processorGroup.shutdownGracefully(0, 10, TimeUnit.SECONDS).sync();

who guarantees you that those threads will live for the 10s they need to gracefully shut down if they're all daemon threads and the JVM will stop without joining them?

Reply:
> if they're all daemon threads and the JVM will stop without joining them?

I don't see why the JVM would stop as we're waiting for the closefuture to return by calling sync. That said I'm not familiar enough with the code to understand the shutdown sequence (those eventloop shutdown methods are a rabbit hole).

Reply:
@jsvd 

> closefuture to return by calling 

ah that's what I was looking for. Makes sense :) That said, since we're doing that, why do we need to make those thread daemon threads in the first place if we ensure their shutdown anyways?

Reply:
I guess my counter point is the reverse of that question: since we're controlling the shutdown sequence, why allow jvm to block us from terminating (which happened in the `exit(1)` issue with jruby)

Reply:
@jsvd 

> why allow jvm to block us from terminating

Fair point, the functional impact here is pretty slim. But imo it's more of a correctness thing:
* If our shutdown handling with the `sync` call and the like works fine => setting `daemon == true` is dead code
* If it's not working properly, we get visibility on that

---------------------

Sorry for wasting time though, I think for HTTP in particular (with its short lived connections) this is very very minor. This is more something to think about in the context of E2E acking/Beats, where it probably makes sense to have that level of formal correctness in this stuff.